pkg/builder/builder.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2019 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package builder
+
+import (
+	"context"
+
+	"github.com/prometheus/client_golang/prometheus"
+	vpaclientset "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/client/clientset/versioned"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	internalstore "k8s.io/kube-state-metrics/internal/store"
+	ksmtypes "k8s.io/kube-state-metrics/pkg/builder/types"
+	"k8s.io/kube-state-metrics/pkg/options"
+)
+
+// Builder helps to build store. It follows the builder pattern
+// (https://en.wikipedia.org/wiki/Builder_pattern).
+type Builder struct {
+	internal ksmtypes.BuilderInterface
+}
+
+// NewBuilder returns a new builder.
+func NewBuilder() *Builder {
+	b := &Builder{
+		internal: internalstore.NewBuilder(),
+	}
+	return b
+}
+
+// WithMetrics sets the metrics property of a Builder.
+func (b *Builder) WithMetrics(r *prometheus.Registry) {
+	b.internal.WithMetrics(r)
+}
+
+// WithEnabledResources sets the enabledResources property of a Builder.
+func (b *Builder) WithEnabledResources(c []string) error {
+	return b.internal.WithEnabledResources(c)
+}
+
+// WithNamespaces sets the namespaces property of a Builder.
+func (b *Builder) WithNamespaces(n options.NamespaceList) {
+	b.internal.WithNamespaces(n)
+}
+
+// WithSharding sets the shard and totalShards property of a Builder.
+func (b *Builder) WithSharding(shard int32, totalShards int) {
+	b.internal.WithSharding(shard, totalShards)
+}
+
+// WithContext sets the ctx property of a Builder.
+func (b *Builder) WithContext(ctx context.Context) {
+	b.internal.WithContext(ctx)
+}
+
+// WithKubeClient sets the kubeClient property of a Builder.
+func (b *Builder) WithKubeClient(c clientset.Interface) {
+	b.internal.WithKubeClient(c)
+}
+
+// WithVPAClient sets the vpaClient property of a Builder so that the verticalpodautoscaler collector can query VPA objects.
+func (b *Builder) WithVPAClient(c vpaclientset.Interface) {
+	b.internal.WithVPAClient(c)
+}
+
+// WithWhiteBlackList configures the white or blacklisted metric to be exposed
+// by the store build by the Builder.
+func (b *Builder) WithWhiteBlackList(l ksmtypes.WhiteBlackLister) {
+	b.internal.WithWhiteBlackList(l)
+}
+
+// WithCustomGenerateStoreFunc configures a constom generate store function
+func (b *Builder) WithCustomGenerateStoreFunc(f ksmtypes.BuildStoreFunc) {
```

Comment:
Why not make the code always call this? And dependency inject even the "normal" store builder? Also seems less likely that we would mess with this and translate to a bit better stability for you :)

Reply:
I tried to follow the same "Builder" configuration pattern that is used for `WithWhiteBlackList()` and other options. Also I tried to not add extra builder configuration step for the "default" use case.

currently the "normal" store builder is set in the `NewBuilder()` function:
https://github.com/kubernetes/kube-state-metrics/blob/ae822f3a2c7678e009e2cc9f32eef687762c5380/internal/store/builder.go#L71

Do you mean that I should remove this default configuration, and expose new function: `DefaultStoreBuilder() ksmtypes.BuildStoreFunc`. 
Then add the step for everyone
```golang
builder.WithCustomGenerateStoreFunc(DefaultStoreBuilder())
```

Reply:
Yes I think just like `WithWhiteBlackList` function it should always be called.

Reply:
Make sense, I have updated the PR accordingly to your comment.